README.rst

```diff
@@ -74,11 +68,167 @@ To install in a virtual environment in your current project:
+
+Commands
+---------
+
+Commands always start with a fixed header. The first entry is always the command number itself encoded in a single byte. The number of subsequent entries in the header will vary by command. The entire header must be sent as a unit so set the characteristic with the full header packet. You can combine multiple commands into a single write as long as the complete header is in the packet.
+
+Paths use ``/`` as a separator and full paths must start with ``/``. Directory paths
+must end with ``/`` when provided as a full path.
+
+All numbers are unsigned.
+
+All values are aligned with respect to the start of the packet.
+
+Status bytes are ``0x01`` for OK and ``0x02`` for error. Other values for error may be used for specific commands.
+
+``0x10`` - Read a file
+++++++++++++++++++++++
+
+Given a full path, returns the full contents of the file.
+
+The header is four fixed entries and a variable length path:
+
+* Command: Single byte. Always ``0x10``.
+* 1 Byte reserved for padding.
+* Path length: 16-bit number encoding the encoded length of the path string.
+* Chunk offset: 32-bit number encoding the offset into the file to start the first chunk.
+* Chunk size: 32-bit number encoding the amount of data that the client can handle in the first reply.
+* Path: UTF-8 encoded string that is *not* null terminated. (We send the length instead.)
+
+The server will respond with:
+* Command: Single byte. Always ``0x11``.
+* Status: Single byte.
+* 2 Bytes reserved for padding.
+* Chunk offset: 32-bit number encoding the offset into the file of this chunk.
+* Total length: 32-bit number encoding the total file length.
+* Chunk length: 32-bit number encoding the length of the read data up to the chunk size provided in the header.
+* Chunk-length contents of the file starting from the current position.
+
+If the chunk length is smaller than the total length, then the client will request more data by sending:
+* Command: Single byte. Always ``0x12``.
+* Status: Single byte. Always OK for now.
+* 2 Bytes reserved for padding.
+* Chunk offset: 32-bit number encoding the offset into the file to start the next chunk.
+* Chunk size: 32-bit number encoding the number of bytes to read. May be different than the original size. Does not need to be limited by the total size.
+
+The transaction is complete after the server has replied with all data. (No acknowledgement needed from the client.)
+
+``0x20`` - Write a file
++++++++++++++++++++++++
+
+Writes the content to the given full path. If the file exists, it will be overwritten. Content may be written as received so an interrupted transfer may lead to a truncated file.
```

Comment:
> If the file exists, it will be overwritten  

Should we have an option to append to existing file, it would be useful for logging.

Reply:
You can do this by setting the offset to the end of the file. You can't "leave it open" though. You'll need to know the next total file size.

Reply:
I'm trying to be pretty specific so that everything is clear between the server and client and also specific about memory requirements.